enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,87 @@
+---
+title: image-content-source-policy-awareness
+authors:
+  - "@sallyom"
+reviewers:
+  - "@smarterclayton"
+  - "@soltysh"
+  - "@wking"
+approvers:
+  - "@smarterclayton"
+creation-date: 2020-05-19
+last-updated: 2020-05-19
+status: provisional
+---
+
+# ImageContentSourcePolicy (ICSP) Awareness
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+There have been several bugs opened around the experience of a 
+user in a disconnected environment using `oc adm release` commands.  If
+using a mirrored image and the mirrored source registry is connected, 
+the following commands do not succeed when in a disconnected environment:
+
+```console
+$ oc adm release extract --tools aprivateregistry/repo/name:tag
+$ oc adm release mirror aprivateregistry/repo/name:tag --to someregistry/repo/name
+$ oc adm must-gather
+```
+
+This is because the mirrored image tags (the individual component images from a payload)
+retain references to the mirrored registry, usually something like 
+`quay.io/openshift-release-dev/ocp-v4.0-art-dev`.  
+
+There needs to be logic in `oc` to look for `ImageContentSourcePolicy` either from a file or from a cluster.
+`oc` should look for `ICSP` in the cluster/current context if connected, and if not, from a file in an expected location,
+probably from the current directory.  `oc` should gather information about RepositoryDigestMirrors from ICSP and use that
+when extracting or mirroring images, or when running must-gather.
+
+Current bugs regarding this Issue:   
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823839
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823143 and also for 4.3, 4.5, 4.6
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+* In what order should `oc` gather info about an image?  Need to order the following:
+    * ICSP from file
+    * ICSP from cluster
+    * image reference from user-given image
+
+* Should oc adm release mirror write an ICSP file to the current directory rather than a print to stdout?
+* Should there be a flag to direct oc to re-order the above 3 choices, to tell oc which information source to use?
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  
+
+## Proposal
+
+* Add logic to `oc adm release` and/or openshift/library-go/pkg/image to become aware of ICSP file
```

Comment:
I want to add a third option and discuss tradeoffs:

1. `oc adm release info` no extra arguments - IF a cluster is available, TRY to find ICSP and if you can't assume none are possible
2. `oc adm release info --icsp-file X` where `X` is a file on disk that has image content policy source options in it (note we don't do anything like this anywhere else, so this is wierd)
3. `oc adm release info --alternate-repository X` where `X` is an image repository that is searched (like ICSP would be) for any image looked up

The first one is fine (as long as the error logic is taken into account). Given a choice between the second and third one, there are times I want both.  As a user of both commands, the third option is simpler and works in more scenarios - I don't need to craft a YAML, I just say

```
oc image info quay.io/openshift-release-dev/ocp-release@sha256:abcdef123456 --alternate-repository myregistry.com/repo/mirror
```

would lookup the digest in the first link and then fall back to the second.  So that's simpler for someone who is used to dealing with sources and destinations.

The counter argument is that for an admin doing this they will be working with ICSP.  So the correct behavior is to do `oc get imagecontentsourcepolicy -o yaml > /foo` and then `oc adm release info --image-content /foo`.  I kind of like this flow for oc adm release info, and i think it's probably ok for `oc image info` too.

Given that, can we define the behavior of 1 and 2 of flags better and then talk about error fallback on 1 (i.e. we can't fail if no ICSP exists).

I can't think of another command that cares about ICSP.

Reply:
Ok, i'll add that to the proposal - option 3  similar to my --force-prefix PR https://github.com/openshift/oc/pull/427  it is simpler because a user knows exactly what they want, they don't need oc to go down a list.  They know they have the images in myrepo/myrelease:tag  and so they should be able to tell oc to only look there. 

The `oc get imagesourcecontentpolicy -o yaml` only works if you have a cluster.  We need to give disconnected users who don't have the cluster yet the ability to extract from a mirrored release - right now, a mirrored release still has the reference to  `quay.io/openshift-release-dev/ocp-release` this is what prohibits `oc adm release extract` from working in disconnected.

Reply:
I've updated this proposal, to reflect my opinion that we need the following:
* A user can specify an ICSP file
* A user can tell oc to use ICSP from cluster
* A user can specify to use image prefix from user-given release image

Reply:
I don't see any reason why we can't support both no 2 and 3 from Clayton's list here. The overlap, so the underlying functionality will be the same, the reading from ICSP or from flag will be the only difference here.

Reply:
With so many flags already for the `oc adm release` commands, I was thinking the 
`--alternative-repository` flag could be avoided.  I will add it, though, if we need it (and since @smarterclayton suggested it, he's probably thinking we do).  I am thinking a user would give their repository and expect to either run the cmd against that exact repository, or from the image reference, or from the ICSP data. These are already there, without the extra flag. 
Example:
`oc adm release extract --tools quay.io/sallyom/test-release:4.5.0-0.nightly-2020-06-22-125424 -v=2`  will extract from quay.io/sallyom/test-release  - this fails from master unless you have the credentials for 
`quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d960fc054c342....`  since that is where the quay.io/sallyom image is mirrored from.  
I don't think a user would run
`oc adm release extract --tools quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d960fc05... --alternative-repository quay.io/sallyom/test-release`  that seems like it's not necessary to me, but I'll add it if we want it.